src/geyser/core.zig

```diff
@@ -0,0 +1,628 @@
+        // make sure we have enough space in the buffer
+        if (expected_n_bytes > self.io_buf.len) {
+            // TODO(metrics): prometheus metrics on io_buf size
+            const new_buf = try self.allocator.alloc(u8, expected_n_bytes);
+            self.allocator.free(self.io_buf);
+            self.io_buf = new_buf;
+        }
+
+        var total_bytes_read: u64 = 0;
+        while (total_bytes_read < expected_n_bytes) {
+            const n_bytes_read = self.file.read(self.io_buf[total_bytes_read..expected_n_bytes]) catch |err| {
+                if (err == std.posix.ReadError.WouldBlock) {
+                    if (self.exit != null and self.exit.?.load(.unordered)) {
+                        return error.PipeBlockedWithExitSignaled;
+                    } else {
+                        // pipe is empty but we dont need to exit, so we try again
+                        // TODO(metrics): prometheus metrics on empty loop count
+                        continue;
+                    }
+                } else {
+                    return err;
+                }
+            };
+
+            if (n_bytes_read == 0) {
+                return error.PipeClosed;
+            }
+            total_bytes_read += n_bytes_read;
+            // TODO(metrics): prometheus metrics on total_bytes_read + expected_n_bytes
+        }
+
+        while (true) {
+            const data = bincode.readFromSlice(
+                self.bincode_allocator.allocator(),
+                T,
+                self.io_buf[0..expected_n_bytes],
+                .{},
+            ) catch |err| {
+                if (err == std.mem.Allocator.Error.OutOfMemory) {
+                    // resize the bincode allocator and try again
+                    const new_size = self.bincode_buf.len * 2;
+                    // TODO(metrics): prometheus metrics on bincode_buf resize
+                    const new_buf = try self.allocator.alloc(u8, new_size);
+                    self.allocator.free(self.bincode_buf);
+                    self.bincode_buf = new_buf;
+                    self.bincode_allocator = std.heap.FixedBufferAllocator.init(self.bincode_buf);
+                    continue;
+                } else {
+                    return err;
+                }
+            };
+
+            return data;
+        }
+    }
+};
+
+pub fn openPipe(pipe_path: []const u8) !std.fs.File {
+    // setup the pipe
+    const rc = c.mkfifo(@ptrCast(pipe_path), c.S_IRWXU);
+    if (rc != 0) {
+        const err = std.posix.errno(rc);
+        switch (err) {
+            // if the pipe already exists, thats ok
+            .EXIST => {},
+            else => {
+                std.debug.print("Failed to create pipe: {}\n", .{err});
+                return error.FailedToCreatePipe;
+            },
+        }
+    }
+
+    const file = try std.fs.cwd().openFile(pipe_path, .{ .mode = .read_write });
+
+    // set to nonblocking
+    // TODO(x19): use F_GETFL | NON_BLOCK
```

Comment:
should this be addressed?

Reply:
its fine for now since we dont expect any preset flags on the pipe - the idea in the TODO was to use `F_GETFL` to get the current fcn flags and then OR in the NON_BLOCK flag so we dont reset any preset flags